### ISSUE TITLE
Erweiterungen des Vorhersage-Algorithmus

### ISSUE THREAD

**Issue by michael-rapp** [mrapp-ke/MLRL-Boomer#360]
Der Algorithmus, der zur Ermittlung von Vorhersagen genutzt wird, erlaubt es aktuell lediglich, die aggregierten Vorhersagescores für jedes Label anhand eines Thresholds in positive/negative Labels umzuwandeln. Je nach verwendeter Konfiguration, bzw. der verwendeten Loss-Funktion, sind aber teilweise andere Vorhergehensweisen sinnvoll bzw. überhaupt möglich.

Um, je nach Konfiguration, die verschiedene Arten von Vorhersagen produzieren zu können, könnte der Vorhersage-Algorithmus um folgende Möglichkeiten erweitert werden.

#### Vorhersage von Wahrscheinlichkeiten

Loss-Funktionen wie z.B. das "Label-wise Logistic Loss" erlauben es, Wahrscheinlichkeiten für jedes Label zu berechnen. In einem solchen Fall wäre es sinnvoll neben Scikit-learns `predict`-Funktion auch die `predict_proba`-Funktion zu implementieren.

* [x] Wenn Label-Wise Logistic Loss verwendet wird
* ~~Bei der Transformation in ein Multiklassen-Problem. Dies erfordert es, dass für jedes Labelset in den Trainingsdaten die Distanz zu den Vorherhersagescores berechnet wird. Anschließend wird für jedes Label die Summe der Distanzen berechnet, die zu Labelsets gehören für die dieses Label gesetzt ist. Die Wahrscheinlichkeit für ein Label berechnet sich dann als `Summe für Label / Summe für alle Labelsets`.~~ Da die Vorhersage von Wahrscheinlichkeiten für mehrere Klassen statt für mehrere Labels ein andersartiges Format erfordert, ist die Umsetzung dieser Funktionalität vorerst nicht geplant.

#### Transformation in Multiklassen-Problem

Loss-Funktionen wie das "Example-wise Logistic Loss` zielen darauf ab das Subset-0/1-Loss zu minimieren, d.h. für jedes Beispiel den korrekten Label-Vektor vorherzusagen. In einem solchen Fall kann es sinnvoll sein, das originale Multilabel-Problem in ein Multiklassen-Problem zu transformieren, bei dem jede Klasse einem Label-Vektor aus den Trainingsdaten entspricht (Subset Correction).

* [x] Umsetzung für Example-wise angewandte Loss-Funktionen,
* [x] Umsetzung für Label-wise angwandte Losses

#### Regression

Der Algorithmus ist aktuell darauf ausgelegt stets Vorhersagen in Form von positiven/negativen Labels zu produzieren. Dies ist auch der Fall wenn Loss-Funktionen, die für Regression konzipiert sind, wie z.B. "Label-wise Squared Error Loss" verwendet werden. In Zukunft könnte die Vorhersage für Regressionsprobleme unterstützt werden, indem unter anderem die folgenden Änderungen umgesetzt werden:

* [x] Es müssen aggregierte Vorhersagescores, ohne jede weitere Transformation, vorhergesagt werden können. 

Scikit-learn trennt strikt zwischen Klassifizierern (sollten die Klasse `sklearn.base.ClassifierMixin` implementieren) und Regressoren (sollten die Klasse `sklearn.base.RegressorMixin` implementieren). Deshalb sollten wohl zwei Varianten (z.B. `BoomerClassifier` und `BommerRegressor`) des bestehenden Algorithmus erstellt werden, die sich lediglich darin unterscheiden, welche Loss-Funktionen verwendet werden können. Außerdem sollten für Regressoren andere Evaluationsmaße verwendet werden. Desweiteren erfordert die Unterstützung von Regressionsproblemen die Möglichkeit, den Trainingsbeispielen beliebige reelle Werte zuzuordnen statt eine binäre Label-Matrix zu verwenden. Da der Zugriff auf die Ground Truth ein wesentlicher Aspekt großer Teile der Implementierung sind und die Unterstützung unterschiedlicher Datenformate dementsprechend umfangreiche Ergänzungen erfordern, geht die Unterstützung reeller Ground Truth-Werte über den hier erfassten Umfang hinaus.
